### DISCUSSION TITLE
create a pre filter for _find queries

### DISCUSSION THREAD

**Discussion by khl3o** [apache/couchdb#4334] [Q&A]
To improve the query time response, I would like to know if it's possible to pre-filter the documents where for a _find query is applied :

Create an index like :

    {
      "index": {
        "partial_filter_selector": {
          "type": {
            "$eq": "installation"
          },
          "contract.country": {
            "$eq": "us"
          }
        }
      },
      "ddoc" : "installation-us",
      "type" : "json"
    }

So if I execute a query like

    {
       "selector": {},
       "use_index": "installation-us"
    }

I will only get the documents where `doc.type = "installation" and doc.contract.country = "us"`. And the possibility to create a selector only on this specific documents (and apply the selector only on 5k instead of 500k documents).

**Comment by willholley** [comment]
I think you will need to include the partial_filter_selector in the
selector passed to _find (in addition to your more discriminative selector)
but otherwise this should work.

**Reply by khl3o** [reply]
thanks for the answer, pre-calculate an index with the filtered documents (like a view) is not possible with couchDB ?

**Reply by willholley** [reply]
@khl3o it is possible using [partial_filter_selector](https://docs.couchdb.org/en/3.2.2-docs/api/database/find.html#partial-indexes). Ignore my previous comment - your example should work, except that indexes require a `"fields"` array which, if you don't care about, you could just set to `["_id"]`:

```
{
  "index": {
    "fields": [
      "_id"
    ],
    "partial_filter_selector": {
      "type": {
        "$eq": "installation"
      },
      "contract.country": {
        "$eq": "us"
      }
    }
  },
  "ddoc" : "installation-us",
  "type" : "json"
}
```
